### ISSUE TITLE
Infrastructure Plan

### ISSUE THREAD

**Issue by WanderingStar** [Data4Democracy/internal-displacement#86]
Here's a sketch of an infrastructure plan:

Development:
Scrapers run locally (on developer machine) in Docker for prototyping (internal-displacement)
Write to Amazon RDS database with shared credential

Web app runs locally in Docker for prototyping (internal-displacement-web)
Reads from Amazon RDS database with shared credential
Can write scrape requests to DB

IDETECT Preparation:
Scraper and Web app Docker containers deployed to AWS instance or similar cloud-hosted infrastructure
Access to Amazon RDS database uses a new credential, access with the old credential is disabled

**Comment by WanderingStar** [comment]
Comments? Suggestions?

**Comment by WanderingStar** [comment]
The purpose of using Docker is to encourage folks to properly encapsulate any dependencies (Texract has a bunch of non-Python dependencies, for example) so that when the time comes to deploy to a cloud instance we don't have to wrangle them.

**Comment by georgerichardson** [comment]
@WanderingStar This looks good to me. By "read scrape requests from the database" do you mean that new additions to the database could automatically trigger scraping?

**Comment by WanderingStar** [comment]
Yes, exactly. The idea would be that we would have a table listing scraping requests and their statuses. The scraper would look for any requests with a NEW status or whatever and scrape them (changing the one it was working on to IN PROGRESS while it worked and COMPLETE or DUPLICATE or whatever when it's done).

This allows us to start more than one scraper if necessary during the initial data ingest and not have them step on each others' toes.

One possible downside to this scheme: If you submit a URL using the web interface and want to wait and see the results, the web server will have to repeatedly check (poll) to see if the request has been completed. This is potentially inefficient if there are a lot of individual URL submissions. But I think that that is not an anticipated scenario right now.

**Comment by WanderingStar** [comment]
Incidentally, this is a fairly common "work queue" pattern.

**Comment by domingohui** [comment]
Hi @WanderingStar thanks for the infra plan suggestion! I don't have much experience in a work flow like this, but I'm just trying to understand it. Does the scraper have to periodically check with the DB for new scrape requests? 

Also, regarding  
> One possible downside to this scheme

do you think we can abstract a backend layer just to communicate with DB? which can be used for both scraper to update the tables, and the frontend webapp to query the result? Specifically to your concern of efficiency, the abstracted backend will just send a response to frontend whenever a request scrape is finished by the scraper and updated on DB. 

This may sound like extra unnecessary work, but another benefit is that both the scraper and the webapp wouldn't be affected if we decided to switch to another DB service. Also, the DB query api can be simplified (i.e SELECT and UPDATE statements won't have to exist in the scraper/webapp anymore.)

What do you think?

**Comment by WanderingStar** [comment]
I strongly recommend using Data Access Objects to talk to the DB, rather than writing SELECT and UPDATE statements. I'm planning to use SQLAlchemy on the Python side. You can see what that usage would look like here: https://github.com/Data4Democracy/internal-displacement/blob/schema/internal_displacement/tests/test_model.py (note that this branch is still work in progress, and not merged yet)

So you create an Article object, and assign its list of Categories, and all of the DB interaction is handled for you.

**Comment by domingohui** [comment]
Thanks @WanderingStar! The reason I brought it up is that I see your [PR](https://github.com/Data4Democracy/internal-displacement/pull/77) is still using explicit SQL commands. But if you are already working on something like sqlalchemy or some sort of DAO, then that's great!

**Comment by WanderingStar** [comment]
Gotcha. #77 was the "get the AWS RDS DB working just like the SQLite DB" PR. There will be a PR forthcoming off of the `schema` branch that uses SQLAlchemy. (see also https://github.com/Data4Democracy/internal-displacement/issues/73)

**Comment by domingohui** [comment]
Just out of curiosity, does this mean the sqlite3 dependency in [SQLAriticleInterface](https://github.com/Data4Democracy/internal-displacement/blob/master/internal_displacement/pipeline.py#L152) will be replaced with SQLalchemy eventually?

**Comment by WanderingStar** [comment]
Sort of. SQLAlchemy uses the low-level drivers like sqlite3 and psycopg2 to talk to SQLite and PostgreSQL databases. You won't have to interact with anything but the model classes in your code.

**Comment by domingohui** [comment]
Sounds good. Just making sure because I didn't see the changes :)
